### PR TITLE
Logging imrovements; scheduled backups

### PR DESCRIPTION
**Pull request by relandboyle**
This PR makes various improvements to server-side logging.

- Logger configuration is moved out of server.ts and into its own file.
- In production mode, logs are output to both stdout and to a local file.
- Log files are rolled daily, the date is included in the filenames, and the files are compressed via Gzip.
- In production mode, a cron job periodically stores the compressed log files to the database.
- Prevent file duplication on the database
- Implement a log retention policy, e.g., "Retain previous thirty days"
- Create an endpoint to download log files
- Create a "back office" UI for downloading logs

Fixes # N/A

## Type of change

- [x] ✨  New feature (non-breaking change which adds functionality)

## How Can this be tested? Testing Plan to review this PR

- [x] Where in the app does this change take place?
server/utils/loggerConfig.ts and loggerScheduledTasks.ts

- [x] What are the steps to test this change?
tl;dr - Just request a demo.

- Manipulate the date code in loggerConfig / loggerProd.appenders.file.pattern to roll the logs every second or every minute.
- Manipulate the cron expressions loggerScheduledTasks / uploadLogs() and evictLogs() to run the tasks frequently and evict files created recently
- Set server/.env/NODE_ENV variable to 'production'
- Create LOGGER_DB_URI variable with 'log_backup' as the db reference

image

What is the expected output?

- In production mode, the server should store INFO level logs in a local file
- The local log file should be rolled daily 
- The current "hot" log file and any other local log files should be persisted to the database daily
- Log files aged more than 30 days should be deleted from the database daily
- Log files from a specified date range can be downloaded via the admin UI

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [x] I have ensured that my pull request title is descriptive.

### CODE REVIEW

**Review by colinvandergraaf — APPROVED** [review]
LGTM!

File: client/features/auth-pages/BackOffice.tsx [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  );
+
+  const startDateMs = startDate.getTime();
+  const endDateMs = endDate.getTime();
+
+  // Calculate the difference in milliseconds
+  const differenceMs = Math.abs(endDateMs - startDateMs); // Use Math.abs to get a positive difference
+  // Convert milliseconds to days
+  const millisecondsPerDay = 1000 * 60 * 60 * 24;
+  const differenceDays = Math.floor(differenceMs / millisecondsPerDay);
```
Comment (kinskiwu) [inline client/features/auth-pages/BackOffice.tsx]: Since you are using `differenceDays` in both the `useEffect` and `downloadLogs`, consider making a util function to improve the readability:

```typescript
const differenceDays = useMemo(() => {
    const startDateMs = startDate.getTime();
    const endDateMs = endDate.getTime();
    const differenceMs = Math.abs(endDateMs - startDateMs);
    return Math.floor(differenceMs / millisecondsPerDay);
  }, [startDate, endDate]);
```

File: client/features/auth-pages/BackOffice.tsx [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    if (differenceDays > 30) {
+      alert('Date range must be less than 30 days');
+    }
+  }, [startDate, endDate, differenceDays]);
+
+  const shortDateString = (date: Date) => {
+    return date.toISOString().split('T')[0].slice(2).split('-').join('');
+  };
+
+  const downloadLogs = async () => {
```
Comment (kinskiwu) [inline client/features/auth-pages/BackOffice.tsx:35]: Since we have an async operation here, I'd highly recommend we incorporate error handling. Also we can use the `axiosClient` to simply the API call.

```typescript
import axiosClient from '@config/axios';

  const downloadLogs = async () => {
    if (differenceDays > 30) {
      alert('Date range must be less than 30 days');
      return;
    }

    try {
      const response = await axiosClient.post(
        '/v1/logs',
        { startDate, endDate },
        { responseType: 'blob' },
      );

      const blob = new Blob([response.data]);
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement('a');
      link.href = url;
      link.setAttribute(
        'download',
        `linkta-logs_${shortDateString(startDate)}_${shortDateString(endDate)}.zip`,
      );
      document.body.appendChild(link);
      link.click();
      link.parentNode?.removeChild(link);
    } catch (error) {
      if (error instanceof Error) {
        alert(`Error: ${error.message}`);
      } else {
        alert('An unexpected error occurred.');
      }
    }
  };
```

File: client/features/auth-pages/BackOffice.tsx [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+          className={`${style.datePicker}`}
+          name='endDate'
+          selected={endDate}
+          onChange={(date) => date && setEndDate(date)}
+        />
+      </label>
+
+      <h4>Days: {differenceDays} (max: 30)</h4>
+
+      <Button
```
Comment (kinskiwu) [inline client/features/auth-pages/BackOffice.tsx:102]: I know we have an alert for warning the user if the `differenceDays > 30`. We could consider disabling the button if the date range limit is exceeded too.

File: client/features/auth-pages/BackOffice.tsx [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+  // Calculate the difference in milliseconds
+  const differenceMs = Math.abs(endDateMs - startDateMs); // Use Math.abs to get a positive difference
+  // Convert milliseconds to days
+  const millisecondsPerDay = 1000 * 60 * 60 * 24;
+  const differenceDays = Math.floor(differenceMs / millisecondsPerDay);
+
+  useEffect(() => {
+    if (startDate >= endDate) {
+      alert('Start date must be before end date');
```
Comment (kinskiwu) [inline client/features/auth-pages/BackOffice.tsx:28]: I'm guessing `alert` is used here given this feature is likely for internal use only.

File: client/features/auth-pages/BackOffice.tsx
Comment (kinskiwu) [inline client/features/auth-pages/BackOffice.tsx:1]: It seems a bit confusing to keep the `BackOffice` in the `auth-pages` folder. Maybe we can create a new folder for this component?

File: server/controllers/logsController.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+      const startDate = startDateString
+        ? new Date(startDateString)
+        : new Date(Date.now() - 30 * 24 * 60 * 60 * 1000); // 30 days ago
+
+      const endDate = endDateString ? new Date(endDateString) : new Date(); // Current date
+
+      // Validate dates
+      if (isNaN(startDate.getTime()) || isNaN(endDate.getTime())) {
+        throw new Error('Invalid startDate or endDate provided.');
```
Comment (kinskiwu) [inline server/controllers/logsController.ts]: We can throw a 400 `BadRequestError` here instead for better UX

File: server/controllers/logsController.ts [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    // delete aged logs from the local directory
+    try {
+      // Read the logs directory
+      const files = await readdirAsync('./logs');
+
+      logger.debug('Evicting log files:', files);
+
+      for (const filename of files) {
+        // skip the hot log file
+        if (filename === 'linkta-server.log') return;
```
Comment (kinskiwu) [inline server/controllers/logsController.ts]: Do you mean to exit the for loop or it was a typo (maybe you meant to use `continue`)?

**Review by kinskiwu — APPROVED** [review]
Thanks for the good work! Looks great overall. Just some minor issues mainly in the backend and some suggestions for best practices.

File: client/features/auth-pages/BackOffice.tsx
Comment (relandboyle) [reply client/features/auth-pages/BackOffice.tsx:1]: I think we need a new round of file and folder organization in the near future. Maybe we can find a time next week.

File: server/controllers/logsController.ts (hunk 6 above)
Comment (relandboyle) [reply server/controllers/logsController.ts]: I think there may have been an earlier revision that was built around a forEach() method, and the return was there to skip the cycle.

File: client/features/auth-pages/BackOffice.tsx (hunk 1 above)
Comment (relandboyle) [reply client/features/auth-pages/BackOffice.tsx]: Done.

File: client/features/auth-pages/BackOffice.tsx (hunk 2 above)
Comment (relandboyle) [reply client/features/auth-pages/BackOffice.tsx:35]: Done.

File: client/features/auth-pages/BackOffice.tsx (hunk 3 above)
Comment (relandboyle) [reply client/features/auth-pages/BackOffice.tsx:102]: Done.

File: client/features/auth-pages/BackOffice.tsx (hunk 4 above)
Comment (relandboyle) [reply client/features/auth-pages/BackOffice.tsx:28]: That's correct. I didn't want to mess with implementing the alert component; at least not right now.

File: server/controllers/logsController.ts (hunk 5 above)
Comment (relandboyle) [reply server/controllers/logsController.ts]: Done.
